### ISSUE TITLE
Feature request : web push notifications

### ISSUE THREAD

**Issue by gauthamzz** [LouisBarranqueiro/hexo-theme-tranquilpeak#387]
What about having something like push notifications, when a new article or something is being posted by someone .then a push notification is being send. There are many services for this

**Comment by LouisBarranqueiro** [comment]
Hi @gauthamzz, I think it's a nice to have feature and I don't have the time to implement it. What about you?

**Comment by gauthamzz** [comment]
I don't know how to implement it , if you can show me the way. Will try my best

**Comment by LouisBarranqueiro** [comment]
I didn't implement a feature like that before. I don't even know if it is possible and if yes, if it can be done easily with a static site.

**Comment by jrbenito** [comment]
Hi guys,

You will need a server (or service) in the middle. Remember that Hexo just generate static site and there is no "action" beyond HTML transfer from the server where you  host your site. That being said, one alternative you can try is to design a deployment plugin for hexo that actually send a message through a push service API. That would be simple to accomplish and your readers only have to subscribe to that push service (ie. Google Cloud Message, [NMA](http://www.notifymyandroid.com/), etc).

Other option: [AWS SNS](https://aws.amazon.com) can be used to implement a push service and/or/mixed email/sms messaging. You will have to setup a queue, push to it on hexo deploy, and design a interface that can subscribe your readers to a list of recipients. I never used it as push service but I do use it to notify me after a success/failure hexo generate on [travis-ci](http://travis-ci.org). This is paid option and I know there are other suppliers already prepared as push service, you just need use their API.

IFTTT integration: Use your RSS feed on a IFTTT applet to trigger a push notification through one of the available services (like notify my droid, etc).

Other idea: Telegram Channel  and a ChatBot (switch Telegram for: skype, facebook messenger, slack or other way a "chatBot" can push messages to). I use this to be notified by my homeTheather system when a new episode of my favorite TV Show is available. To accomplish this I wrote a microservice that receive a HTTP request and send it as message to me (or family group chat in my case) through  a Telegram ChatBot. Telegram offers a thing called Channel, it is like a whatzapp group chat but only Channel owner can write message while all other members can just receive that message. (very similar to a Push notification). The microservice I wrote is way too simple written (badly) in NodeJS, runs for free on [Hook.io](http://hook.io) but could be run on AWS Lambda or Google Cloud Function, App Engine, Heroku or any other, few modifications needed.

Based on the microservice idea above and a mix of push service or IFTTT I think it should be pretty simple to have a deploy plugin for hexo where the deploy action is a simple API call or a HTTP request to specified URL and PUSH messages would be sent way. The worst part, IMHO, would be to elect a push service  free for the subscribers (your reader) and with no need to install apps (many would no use if they have to install a new app just because of you).

**Comment by gauthamzz** [comment]
what about integrating with services such as [one signal](https://onesignal.com/) and [pushcrew](https://pushcrew.com/)

**Comment by jrbenito** [comment]
@gauthamzz ,

I believe same principle applies. I do not know deeply services you mentioned but for web notification I think you should ad some sort of script to your HTML, this can be done by theme during site generation (just like it adds Google Analytics). However to push messages one would have to login to service and generate push message or, by a "deploy plugin" as earlier suggested, use the service API to push every time a new post is published.

Did you try to integrate some sort of web push from these services?

**Comment by jrbenito** [comment]
From the [OneSignal](https://documentation.onesignal.com/docs/web-push-setup):

> Web push notifications are messages that come from a website. You get them on your desktop or device even when the website is not open in your browser. It is a new marketing channel to re engage your site visitors without knowing their email or other contact details.

Worth mention: (even it seems obvious)
> Note: Incognito / Private Browsing mode isn't supported in browsers.

Reading the guide for [HTTPS](https://documentation.onesignal.com/docs/web-push-sdk-setup-https) sites (there is another for [HTTP](https://documentation.onesignal.com/docs/web-push-sdk-setup-http)), it works exactly as I thought  it would (my previous message). You need to and their scripts  files on your source folder to be copied to public folder during `hexo generate` and modify Tranquilpeak to include the script call below at the head of you site.

```
<link rel="manifest" href="/manifest.json">
  <script src="https://cdn.onesignal.com/sdks/OneSignalSDK.js" async></script>
  <script>
    var OneSignal = window.OneSignal || [];
    OneSignal.push(["init", {
      appId: "YOUR_APP_ID",
      autoRegister: false,
      notifyButton: {
        enable: true /* Set to false to hide */
      }
    }]);
  </script>
```

Taking Google Analytics as example, you shall create file `tranquilpeak/layout/_partial/onesignal.ejs` similar to the `tranquilpeak/layout/_partial/google-analytics.ejs` with content above (taken from the HTTPS guide). And modify  `tranquilpeak/layout/_partial/head.ejs` to include your partial, similar to:

```
165    <!--STYLES END-->
166     <%- partial('google-analytics') %>
167     <%- partial('baidu-analytics') %>
168     <%- partial('onesignal.ejs') %>
169 </head>
```
(Numbers are line numbers on [latest dev branch file version](https://github.com/LouisBarranqueiro/hexo-theme-tranquilpeak/blob/master/layout/_partial/head.ejs#L165-L168), line 168 was introduced here)

After that, your viewer will be presented with a choice to allow or not push notifications to be received and every push you send (i.e. using one signal [dashboard](https://documentation.onesignal.com/docs/dashboard)) will show up in viewer browser.

Now, how to send notification automatically when generating site? Use the [API](https://documentation.onesignal.com/reference#create-notification) POST method. Mount a URL based on parameters you desire and make a call to it (i.e. `curl URL`) after every `hexo generate`. Optionally  a deploy plugin can be created but I think it is overkill, simply write a script that call hexo generate, hexo deploy and curl or use something like [Travis-ci.org](https://travis-ci.org).

If this is interesting to add to Tranquilpeak I can PR it after testing. @LouisBarranqueiro  what do you think?

**Comment by jrbenito** [comment]
@gauthamzz  and @LouisBarranqueiro ,

I did a quick and dirty test by implementing the above and it worked. Iwill organize it now:

1.  Make configurations at _config.yml (done)
2.  There is difference for HTTPS and HTTP sites, now I handle both at .ejs template but not add any of HTTPS javascripts files to assets. For now this might be done at hexo level but will try to do it at theme level during theme build. (partial)
3.  Original push subscribe bell float over page on right down corner by default, need to figure out how to fix it on side bar and/or near share buttons.
4.  Documentantion

At this point it would be ready for PR. Meanwhile, if anyone want to give it try, take a look at [my branch](https://github.com/jrbenito/hexo-theme-tranquilpeak/tree/onesignal-service-integration). 

_help on is welcome_

Please, profide feedback

**Comment by jrbenito** [comment]
Hi,

I worked a little more on [the branch](https://github.com/jrbenito/hexo-theme-tranquilpeak/tree/onesignal-service-integration) and here is the update:

1.  Make configurations at _config.yml _**(done)**_
2.  There is difference for HTTPS and HTTP sites, now I handle both at .ejs template but not add any of HTTPS javascripts files to assets. For now this might be done at hexo level but will try to do it at theme level during theme build. Or a plugin for hexo might download and handle scripts since they need to be at root level... Ideas welcome. **_(partially done)_**
3.  Created a link on side bar showing a model (like about), this modal asks if user accept to subscribe push notifications. Modal still need some makeup at css (help needed) and now it has a link to accept/cancel subscription. I hope to transfome that link in a HTML button styled after tranquilpeak. @LouisBarranqueiro, could you help me on this? **_(partially done)_**
4.  Documentantion/Translations **_(partially done)_**

I PR it just to be visible for anyone visiting this repo, howeve it still need a bit of work to deserve integration in dev branch.

Please comment

**Comment by LouisBarranqueiro** [comment]
Great work @jrbenito ! :)
Could you explain me how do it works? for a reader? and for the blog owner? 
When and how do you notify readers that there is a new post?

**Comment by jrbenito** [comment]
@LouisBarranqueiro ,

**For the reader**, if she clicks on the sidebar icon for push, she will be presented to a floating card (like #about) asking permission to send her push messages. If reader accepts it, reader´s browser is anonymously added to blog´s Owner pool of users at Onesignal.com application dashboard. From reader´s perspective, that is it! Browser will present she with messages anytime blog´s owner sent it.

**From blog´s ownser** perspective, he can manage their push through Onesignal.com app dashboard. He can manually create and send pushes there, automate some things like creating message templates and schedule (I did not explore it deeply), see how many subscribers he got to his blog (readers that opted in for pushes). At this point I did not included any way to send pushes directly from `hexo generate` but this shall be one `HTTP` call to an API and could be done by `curl` or similar or even a small deploy plugin for hexo. This feature is next step. Now, blog owner must go to onesignal.com and send pushes manually or setup his own way to push automatically.

### Thoughts on automatic push: 

**Ideal automation**: The ideal way to push, IMHO, would be to send pushes and deploy stage but keeping track of posts so customized message could be crafted. Something like:

> Hey, come to read my new article "Ingredients for a good cake" I just posted to the blog.

And, of course, the post title is a link directly to the blog post. But keeping track of posts are very important to not send out duplicate messages under risk of users opt out your service. I don´t know how to it yet but think it could be done with deploy plugin and database (same used by hexo maybe).

Anyway, the simple solution is to trigger generic messages linking to the blog´s land page. Like:

> Hey, we have new posts to the blog.

This is much simpler to do.

For test purposes you can subscribe at my [test stage](teste.benito.com.br.s3-website-us-east-1.amazonaws.com).

**Comment by jrbenito** [comment]
Hi @LouisBarranqueiro  and @gauthamzz ,

To make things easy I wrote a very simple deployer plugin for hexo. Its only task now is to take a template file, put your app_id, prepare https headers (authtentication token) and send a request to Onesignal API server. With  this plugin one can do:

```
$ hexo generate
$ hexo deploy
```
And have the template message sent to all users subscribed. The template can have multiple languages and target segments (default template has only english and portuguese messages and targets ALL users, edit it for your needs). In order to use:

`$ npm install --save hexo-deployer-onesignal`

and edit hexo `_config.yml` to add:

```
deploy: 
    -type: onesignal
     app_id: <uuid>
```
and export your authorization token in a environment variable:

```
$ export ONESIGNAL_APP_AUTH_KEY="auth-token-from-the-app"
$ hexo generate
$ hexo deploy
```

The idea is to expand it to read templates from a user defined folder (or just `hexo_dir/templates`) and maybe include some customization logic to select between templates based on the post category or tags (just wondering here).

I really hope this helps @gauthamzz and others.

best regards,
JrBenito.

**Comment by jrbenito** [comment]
add documentation to npmjs.org and github mentioning this theme

**Comment by LouisBarranqueiro** [comment]
You did a great job but I don't want to include all this code directly in the theme because it's justly a theme. All this code should stand in a separate npm module and integrated in the theme with only:
```
<%- partial('hexo-onesignal/lib/templates/notify', {lang: page.lang}) %>
```
All hexo themes should be able to use your work like this. Create a node module, and notify me if you have some questions to achieve this. You should only expose the css variable names to able each theme to customize style of the modal, that's all.

To sum up, your node module `hexo-onesignal` should includes:
- the template of the modal
- the script to animate the modal
- the css to stylize the modal (each theme should be able to override the style to customize the modal)
- the translations for the modal
- the deployer to notify user when a new version is deployed

**Comment by jrbenito** [comment]
Hi @LouisBarranqueiro,

Thanks for feedback and insights. This started out of curiosity seeded by @gauthamzz request. I never thought it would come to this point ;)

Well, I found your suggestion very interesting although I never realized one can include snippets to a theme like you suggest. I do have doubt on: 

> the css to stylize the modal (each theme should be able to override the style to customize the modal)

You said "expose css" on the PR page, how exactly to achieve this?

Will think about it, sound like good side project.

Thanks
